misk/src/main/kotlin/misk/web/resources/StaticResourceAction.kt

```diff
@@ -0,0 +1,142 @@
+ * ```
+ */
+
+@Singleton
+class StaticResourceAction : WebAction {
+  @Inject private lateinit var entries: List<StaticResourceEntry>
+  @Inject @JvmSuppressWildcards private lateinit var clientRequest: ActionScoped<Request>
+  @Inject private lateinit var resourceLoader: ResourceLoader
+
+  @Get("/{path:.*}")
+  @Post("/{path:.*}")
+  @RequestContentType(MediaTypes.ALL)
+  @ResponseContentType(MediaTypes.ALL)
+  @Unauthenticated  // TODO(adrw) should this be unauthenticated?
+  fun action(): Response<ResponseBody> {
+    val request = clientRequest.get()
+    return getResponse(request)
+  }
+
+  private fun getResponse(request: Request): Response<ResponseBody> {
+    val urlPath = request.url.encodedPath()
+    val matchedEntry =
+        ResourceEntryCommon.findEntryFromUrl(entries, request.url) as StaticResourceEntry?
+    return when (exists(normalizePath(urlPath))) {
+      Kind.NO_MATCH -> when {
+        !urlPath.contains(".") && !urlPath.endsWith("/") ->
+          redirectResponse(normalizePathWithQuery(request.url))
+      // actually return the resource, don't redirect. Path must stay the same since this will be handled by React router
+        !urlPath.contains(".") && urlPath.endsWith("/") ->
+          resourceResponse(normalizePath(matchedEntry?.url_path_prefix ?: urlPath))
+        else -> null
+      }
+      Kind.RESOURCE -> resourceResponse(normalizePath(urlPath))
+      Kind.RESOURCE_DIRECTORY -> redirectResponse(normalizePathWithQuery(request.url))
+    } ?: NotFoundAction.response(request.url.toString())
+  }
+
+  /** Returns true if the mapped path exists on either the resource path or file system. */
+  private fun exists(urlPath: String): Kind {
+    val staticResource =
+        ResourceEntryCommon.findEntryFromUrl(entries, urlPath) as StaticResourceEntry?
+            ?: return Kind.NO_MATCH
+    val resourcePath = staticResource.resourcePath(urlPath)
+    if (resourceLoader.exists(resourcePath)) return Kind.RESOURCE
+    if (resourceLoader.list(resourcePath).isNotEmpty()) return Kind.RESOURCE_DIRECTORY
+    return Kind.NO_MATCH
+  }
+
+  enum class Kind {
+    NO_MATCH,
+    RESOURCE,
+    RESOURCE_DIRECTORY,
+  }
+
+  /** Returns a source to the mapped path, or null if it doesn't exist. */
+  fun open(urlPath: String): BufferedSource? {
+    val staticResource =
+        ResourceEntryCommon.findEntryFromUrl(entries, urlPath) as StaticResourceEntry?
+            ?: return null
+    val resourcePath = staticResource.resourcePath(urlPath)
+    return when {
+      resourceLoader.exists(resourcePath) -> resourceLoader.open(resourcePath)!!
+      else -> null
+    }
+  }
+
+  private fun normalizePath(urlPath: String): String {
+    return when {
+      urlPath.endsWith("/") -> "${urlPath}index.html"
+      !urlPath.contains(".") && !urlPath.endsWith("/") -> "$urlPath/"
+      else -> urlPath
+    }
+  }
+
+  private fun normalizePathWithQuery(url: HttpUrl): String {
```

Comment:
I think that if there’s a query we should probably 404 the request? we don’t want these paths to have queries, that’s unexpected

Reply:
Offline: reviewed that the query gets passed on to the React frontend so we'll leave this in.